tensorpipe/channel/cuda_basic/channel.cc

```diff
@@ -0,0 +1,358 @@
+  TP_VLOG(4) << "Channel " << id_ << " received a send request (#"
+             << sequenceNumber << ")";
+
+  if (error_) {
+    descriptorCallback(error_, std::string());
+    callback(error_);
+    return;
+  }
+
+  TP_VLOG(4) << "Channel " << id_
+             << " is copying buffer from CUDA device to CPU";
+  auto tmpBuffer = makeCudaPinnedBuffer(buffer.length);
+  TP_CUDA_CHECK(cudaMemcpyAsync(
+      tmpBuffer.get(),
+      buffer.ptr,
+      buffer.length,
+      cudaMemcpyDeviceToHost,
+      buffer.stream));
+
+  auto cudaCallback =
+      [this,
+       buffer,
+       tmpBuffer{std::move(tmpBuffer)},
+       descriptorCallback{std::move(descriptorCallback)},
+       callback{std::move(callback)}](cudaError_t cudaError) mutable {
+        TP_VLOG(4) << "Channel " << id_
+                   << " is done copying buffer from CUDA device to CPU";
+        TP_CUDA_CHECK(cudaError);
+
+        // TODO: Use eagerCallbackWrapper to ensure channel remains alive until
+        // this callback fires?
+
+        loop_.deferToLoop([this,
+                           buffer,
+                           tmpBuffer{std::move(tmpBuffer)},
+                           descriptorCallback{std::move(descriptorCallback)},
+                           callback{std::move(callback)}]() mutable {
+          if (error_) {
+            descriptorCallback(error_, std::string());
+            callback(error_);
+            return;
+          }
+
+          CpuBuffer cpuBuffer{tmpBuffer.get(), buffer.length};
+          // Keep tmpBuffer alive until cpuChannel_ is done sending it over.
+          callback = eagerCallbackWrapper_(
+              [tmpBuffer{std::move(tmpBuffer)}, callback{std::move(callback)}](
+                  Impl& impl) { callback(impl.error_); });
+          cpuChannel_->send(
+              cpuBuffer, std::move(descriptorCallback), std::move(callback));
+        });
+      };
+  cudaLoop_->addCallback(buffer.stream, std::move(cudaCallback));
+}
+
+// Receive memory region from peer.
+void Channel::recv(
+    TDescriptor descriptor,
+    CudaBuffer buffer,
+    TRecvCallback callback) {
+  impl_->recv(std::move(descriptor), buffer, std::move(callback));
+}
+
+void Channel::Impl::recv(
+    TDescriptor descriptor,
+    CudaBuffer buffer,
+    TRecvCallback callback) {
+  loop_.deferToLoop([this,
+                     descriptor{std::move(descriptor)},
+                     buffer,
+                     callback{std::move(callback)}]() mutable {
+    recvFromLoop(std::move(descriptor), buffer, std::move(callback));
+  });
+}
+
+void Channel::Impl::recvFromLoop(
+    TDescriptor descriptor,
+    CudaBuffer buffer,
+    TRecvCallback callback) {
+  TP_DCHECK(loop_.inLoop());
+
+  const uint64_t sequenceNumber = nextTensorBeingReceived_++;
+  TP_VLOG(4) << "Channel " << id_ << " received a recv request (#"
+             << sequenceNumber << ")";
+
+  if (error_) {
+    callback(error_);
+    return;
+  }
+
+  TP_DCHECK_EQ(descriptor, std::string());
```

Comment:
Why would it be empty? It's the CPU channel's descriptor, it could be anything right?